src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5249,21 +5249,18 @@ public static Task<TResult> FromResult<TResult>(TResult result)
                         return Unsafe.As<Task<TResult>>(task); // Unsafe.As avoids a type check we know will succeed
                     }
                 }
-                // For other known value types, we only special-case 0 / default(TResult).  We don't include
-                // floating point types as == may return true for different bit representations of the same value.
-                else if (
-                    (typeof(TResult) == typeof(uint) && default == (uint)(object)result!) ||
-                    (typeof(TResult) == typeof(byte) && default(byte) == (byte)(object)result!) ||
-                    (typeof(TResult) == typeof(sbyte) && default(sbyte) == (sbyte)(object)result!) ||
-                    (typeof(TResult) == typeof(char) && default(char) == (char)(object)result!) ||
-                    (typeof(TResult) == typeof(long) && default == (long)(object)result!) ||
-                    (typeof(TResult) == typeof(ulong) && default == (ulong)(object)result!) ||
-                    (typeof(TResult) == typeof(short) && default(short) == (short)(object)result!) ||
-                    (typeof(TResult) == typeof(ushort) && default(ushort) == (ushort)(object)result!) ||
-                    (typeof(TResult) == typeof(IntPtr) && default == (IntPtr)(object)result!) ||
-                    (typeof(TResult) == typeof(UIntPtr) && default == (UIntPtr)(object)result!))
+                // For other value types, we special-case default(TResult) if we can easily compare bit patterns to default/0.
+                else if (!RuntimeHelpers.IsReferenceOrContainsReferences<TResult>())
                 {
-                    return Task<TResult>.s_defaultResultTask;
+                    // We don't need to go through the equality operator of the TResult because we cached a task for default(TResult),
+                    // so we only need to confirm that this TResult has the same bits as default(TResult).
```

Comment:
This has unpredictable behavior for structs that are not tightly packed. The uninitialized hole in such structs may cause the cache to kick in intermittently. It should not affect correctness though.

Reply:
Is any code going to do reference equal on what's returned here?

Reply:
We have already warned that ["for some `T` types and some result values, `FromResult` may return a cached singleton object"](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/task-fromresult-returns-singleton#new-behavior). This change merely broadens the conditions under which a cached singleton will be returned.

Reply:
> This has unpredictable behavior for structs that are not tightly packed.

Yeah, I considered that and adding an is tightly packed check, but then thought of it from the opposite persistive: by allowing such types, we're opportunistically able to avoid more allocations whereas if it's banned we'd never avoid them.

Reply:
> We have already warned

Yup. And even then I suspect almost no one was negatively impacted, I was just being extra cautious with that breaking change announcement. I've never seen anyone rely on reference equality on the result of FromResult; most devs I've talked to about it over the years already thought it was caching.